crates/bevy_pbr/src/render_graph/forward_pipeline/forward.frag

```diff
@@ -24,41 +27,170 @@ layout(set = 1, binding = 0) uniform Lights {
+    return baseColor * (1.0 - metallic);
+}
+
+#define MIN_N_DOT_V 1e-4
+
+float clampNoV(float NoV) {
+    // Neubelt and Pettineo 2013, "Crafting a Next-gen Material Pipeline for The
+    // Order: 1886"
+    return max(NoV, MIN_N_DOT_V);
+}
+
+float perceptualRoughnessToRoughness(float perceptualRoughness) {
+    return perceptualRoughness * perceptualRoughness;
+}
+
+#if defined(TARGET_MOBILE)
+// min roughness such that (MIN_PERCEPTUAL_ROUGHNESS^4) > 0 in fp16 (i.e.
+// 2^(-14/4), rounded up)
+#define MIN_PERCEPTUAL_ROUGHNESS 0.089
+#define MIN_ROUGHNESS 0.007921
+#else
+#define MIN_PERCEPTUAL_ROUGHNESS 0.045
+#define MIN_ROUGHNESS 0.002025
+#endif
+
+#endif
 
 void main() {
     vec4 output_color = Albedo;
-# ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
-    output_color *= texture(
-        sampler2D(StandardMaterial_albedo_texture, StandardMaterial_albedo_texture_sampler),
-        v_Uv);
-# endif
-
-# ifndef STANDARDMATERIAL_UNLIT
-    vec3 normal = normalize(v_Normal);
+#ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
+    output_color *= texture(sampler2D(StandardMaterial_albedo_texture,
+                                      StandardMaterial_albedo_texture_sampler),
+                            v_Uv);
+#endif
+
+#ifndef STANDARDMATERIAL_UNLIT
+    float perceptual_roughness = pbr.x;
+    perceptual_roughness =
+        clamp(perceptual_roughness, MIN_PERCEPTUAL_ROUGHNESS, 1.0);
+    float roughness = perceptualRoughnessToRoughness(perceptual_roughness);
+
+    float metallic = pbr.y;
+    vec3 N = normalize(v_Normal);
+    vec3 V = normalize(CameraPos.xyz - v_Position.xyz);
+
+    vec3 F0 = vec3(0.04);
+    F0 = mix(F0, output_color.rgb, metallic);
+
+    vec3 diffuseColor = computeDiffuseColor(output_color.rgb, metallic);
+
     // accumulate color
-    vec3 color = AmbientColor;
-    for (int i=0; i<int(NumLights.x) && i<MAX_LIGHTS; ++i) {
+    vec3 light_accum = vec3(0.0);
+    for (int i = 0; i < int(NumLights.x) && i < MAX_LIGHTS; ++i) {
         Light light = SceneLights[i];
-        // compute Lambertian diffuse term
-        vec3 light_dir = normalize(light.pos.xyz - v_Position);
-        float diffuse = max(0.0, dot(normal, light_dir));
-        // add light contribution
-        color += diffuse * light.color.xyz;
+
+        vec3 lightDir = light.pos.xyz - v_Position.xyz;
+        vec3 L = normalize(lightDir);
+
+        float rangeAttenuation =
+            getDistanceAttenuation(lightDir, light.attenuation);
+
+        vec3 H = normalize(L + V);
+
+        float NdotL = clampNoV(dot(N, L));
+        float NdotV = clampNoV(dot(N, V));
+        float NoL = saturate(NdotL);
+        float NoH = saturate(dot(N, H));
+        float LoH = saturate(dot(L, H));
+
+        vec3 specular = isotropicLobe(F0, roughness, H, NdotV, NoL, NoH, LoH);
+        vec3 diffuse = diffuseColor * Fd_Burley(roughness, NdotV, NoL, LoH);
+
+        light_accum +=
+            ((diffuse + specular) * light.color.xyz) * (light.color.w * NdotL);
```

Comment:
Unless you're going to multiply the incoming light intensity by PI, you'd need to do that here so that the light intensities behave the way artists expect. Also, you could just remove the `1 / PI` parts from the equations above and just leave a comment in them saying they'd be cancelled out anyway.
https://seblagarde.wordpress.com/2012/01/08/pi-or-not-to-pi-in-game-lighting-equation/

Reply:
Are you on the bevy discord, by any chance? Would make it easier to ask some questions that don't really need to be archived here. My nick is the same there.

Reply:
Do you mean `1.0 / PI` here? See https://google.github.io/filament/Filament.html#mjx-eqn-pointLightLuminanceEquation

Reply:
I've added what I think is the physically correct term in https://github.com/bevyengine/bevy/pull/1554/commits/ac1e23c83a1c9576b82263dea3f1bd845bb9a46e

I'll read the linked article and see if I can figure it out.

Reply:
I can't find anything like this term PI in either the filament document or the code and my math is not good enough to reproduce the derivations.

Reply:
I've got a better grasp on this now, but I'd love for somebody else to chime in on how we want the light to behave. I agree that if we end up multiplying by `PI` we might as well remove the `1.0 / PI` instead.